### PR TITLE
[23926] - Store & Show Challenges from Notification Extension.

### PR DESCRIPTION
**Pull request by AlejandroOrozco**
## Ticket
- [23926]

## Description
- Store & Show Challenges from Notification Extension

## Testing
- [ ] Added unit tests
- [ ] Ran unit tests successfully
- [ ] Added documentation for public APIs and/or Wiki

### CODE REVIEW

File: TwilioVerifyDemo/TwilioVerifyDemo/TwilioVerify/TwilioVerifyAdapter.swift [hunk 1]
```diff
@@ -27,7 +27,9 @@ class TwilioVerifyAdapter {
     #if DEBUG
       builder = builder.enableDefaultLoggingService(withLevel: .all)
     #endif
-    twilioVerify = try builder.build()
+    twilioVerify = try builder
+      .setAccessGroup("group.company.identifier")
```
Comment (yafuquen) [inline TwilioVerifyDemo/TwilioVerifyDemo/TwilioVerify/TwilioVerifyAdapter.swift:31]: We could add a TODO for setting the customer's correct access group. Should we use the access group for the sample app in order to generate builds for testing?

Comment (yafuquen) [reply TwilioVerifyDemo/TwilioVerifyDemo/TwilioVerify/TwilioVerifyAdapter.swift:31]: And we can include also where to update the access group for the extension

File: TwilioVerifyDemoCache/Tests/TwilioVerifyDemoCacheTests.swift [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import XCTest
+@testable import TwilioVerifyDemoCache
+
+final class TwilioVerifyDemoCacheTests: XCTestCase {
```
Comment (yafuquen) [inline TwilioVerifyDemoCache/Tests/TwilioVerifyDemoCacheTests.swift:4]: Should we need this?

File: TwilioVerifySDK/TwilioSecurity/Sources/Key/Template/ECP256SignerTemplate.swift [hunk 3]
```diff
@@ -27,17 +27,28 @@ public struct ECP256SignerTemplate {
   public var algorithm = kSecAttrKeyTypeECSECPrimeRandom as String
   public var signatureAlgorithm = SecKeyAlgorithm.ecdsaSignatureMessageX962SHA256
   public var accessControl: SecAccessControl
+  public var accessGroup: String?
```
Comment (yafuquen) [inline TwilioVerifySDK/TwilioSecurity/Sources/Key/Template/ECP256SignerTemplate.swift:30]: Please update the base branch to not see the changes for the SDK

File: TwilioVerifySDK/TwilioVerify/Sources/Domain/Challenge/PushChallengeProcessor.swift [hunk 4]
```diff
@@ -29,32 +29,58 @@ class PushChallengeProcessor {
...
+    failure: @escaping TwilioVerifyErrorBlock
+  ) {
     Logger.shared.log(withLevel: .info, message: "Getting challenge \(sid) with factor \(factor.sid)")
     challengeProvider.get(withSid: sid, withFactor: factor, success: success, failure: { error in
       failure(TwilioVerifyError.inputError(error: error as NSError))
     })
   }
   
-  func updateChallenge(withSid sid: String, withFactor factor: PushFactor, status: ChallengeStatus, success: @escaping EmptySuccessBlock, failure: @escaping TwilioVerifyErrorBlock) {
+  func updateChallenge(
```
Comment (AuthyApps) [inline TwilioVerifySDK/TwilioVerify/Sources/Domain/Challenge/PushChallengeProcessor.swift:54]: <table data-meta="generated_by_UnitTests">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-ios/blob/c5b2840754ffa41023a18cddc2098838d657bd0d/TwilioVerifySDK/TwilioVerify/Sources/Domain/Challenge/PushChallengeProcessor.swift#L54"></span>Function should have complexity 10 or less: currently complexity equals 11<br />
<code>cyclomatic_complexity</code> <code>PushChallengeProcessor.swift:54</code></td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline TwilioVerifySDK/TwilioVerify/Sources/Domain/Challenge/PushChallengeProcessor.swift:54]: <table data-meta="generated_by_UnitTests">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-ios/blob/c5b2840754ffa41023a18cddc2098838d657bd0d/TwilioVerifySDK/TwilioVerify/Sources/Domain/Challenge/PushChallengeProcessor.swift#L54"></span>Function body should span 40 lines or less excluding comments and whitespace: currently spans 60 lines<br />
<code>function_body_length</code> <code>PushChallengeProcessor.swift:54</code></td>
    </tr>
  </tbody>
</table>

**Comment by AuthyApps** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

## Code coverage
### Total coverage: **`93.65%`**
File | Coverage
-----|-----
TwilioVerifySDK/TwilioSecurity/Sources/Key/Signer/ECSigner.swift | **`94.74%`**
TwilioVerifySDK/TwilioSecurity/Sources/Key/Template/ECP256SignerTemplate.swift | **`92.11%`**
TwilioVerifySDK/TwilioSecurity/Sources/Keychain/KeyManager.swift | **`85.71%`**
TwilioVerifySDK/TwilioSecurity/Sources/Keychain/Keychain.swift | **`92.45%`**
TwilioVerifySDK/TwilioSecurity/Sources/Keychain/KeychainQuery.swift | **`100.00%`**
TwilioVerifySDK/TwilioSecurity/Sources/Storage/SecureStorage.swift | **`84.62%`**
TwilioVerifySDK/TwilioVerify/Sources/Data/JWT/JwtGenerator.swift | **`100.00%`**
TwilioVerifySDK/TwilioVerify/Sources/Data/JWT/JwtSigner.swift | **`75.76%`**
TwilioVerifySDK/TwilioVerify/Sources/Data/KeyStorage.swift | **`92.50%`**
TwilioVerifySDK/TwilioVerify/Sources/Data/Storage+AccessGroup.swift | **`93.33%`**
TwilioVerifySDK/TwilioVerify/Sources/Data/Storage.swift | **`90.84%`**
TwilioVerifySDK/TwilioVerify/Sources/Data/StorageProvider.swift | **`80.85%`**
TwilioVerifySDK/TwilioVerify/Sources/Domain/Challenge/ChallengeFacade.swift | **`96.63%`**
TwilioVerifySDK/TwilioVerify/Sources/Domain/Challenge/ChallengeRepository.swift | **`94.12%`**
TwilioVerifySDK/TwilioVerify/Sources/Domain/Challenge/PushChallengeProcessor.swift | **`96.91%`**
TwilioVerifySDK/TwilioVerify/Sources/Domain/Factor/FactorFacade.swift | **`97.60%`**
TwilioVerifySDK/TwilioVerify/Sources/Domain/Factor/PushFactory.swift | **`98.83%`**
TwilioVerifySDK/TwilioVerify/Sources/Domain/Manager/TwilioVerify.swift | **`79.41%`**
TwilioVerifySDK/TwilioVerify/Sources/Networking/Authentication.swift | **`100.00%`**
TwilioVerifySDK/TwilioVerify/Sources/TwilioVerifyError.swift | **`94.32%`**
> Powered by [Slather](https://github.com/SlatherOrg/slather)

<p align="right" data-meta="generated_by_UnitTests">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by mejiagarcia — APPROVED** [review]

**Comment by AuthyApps** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

## Code coverage
### Total coverage: **`93.61%`**
> Powered by [Slather](https://github.com/SlatherOrg/slather)

<p align="right" data-meta="generated_by_UnitTests">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: TwilioVerifyDemo/NotificationExtension/NotificationService.swift [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+final class NotificationService: UNNotificationServiceExtension {
+
+  // MARK: - Properties
+  
+  var contentHandler: ((UNNotificationContent) -> Void)?
+  var bestAttemptContent: UNMutableNotificationContent?
+
+  private lazy var twilioVerify: TwilioVerify? = try? TwilioVerifyBuilder()
+    .setAccessGroup("group.company.identifier")
```
[the remark this answers is not in the document]
Comment (AlejandroOrozco) [reply TwilioVerifyDemo/NotificationExtension/NotificationService.swift]: Added  `#warning` instead, that would be more visible.

File: TwilioVerifyDemo/TwilioVerifyDemo/TwilioVerify/TwilioVerifyAdapter.swift (hunk 1 above)
Comment (AlejandroOrozco) [reply TwilioVerifyDemo/TwilioVerifyDemo/TwilioVerify/TwilioVerifyAdapter.swift]: Added #warning instead, that would be more visible. Also mentioned the NotificationExtension.

**Review by yafuquen — APPROVED** [review]

**Pull request merged**